fl4health/clients/ditto_mkmmd_client.py

```diff
@@ -0,0 +1,483 @@
+        self.optimizers["local"].zero_grad()
+
+        # Forward pass on both the global and local models
+        preds, features = self.predict(input)
+
+        # Compute all relevant losses
+        losses = self.compute_training_loss(preds, features, target)
+
+        # Take a step with the global model vanilla loss
+        losses.additional_losses["global_loss"].backward()
+        self.optimizers["global"].step()
+
+        # Take a step with the local model using the local loss and Ditto constraint
+        losses.backward["backward"].backward()
+        self.optimizers["local"].step()
+
+        # Return dictionary of predictions where key is used to name respective MetricMeters
+        return losses, preds
+
+    def update_after_step(self, step: int) -> None:
+        if step % self.beta_update_interval == 0:
+            if self.mkmmd_loss_weight and self.init_global_model:
+                # Get the feature distribution of the local and init global features with evaluation mode
+                local_distribution, init_global_distribution = self.update_buffers(self.model, self.init_global_model)
+                # Update betas for the MK-MMD loss based on gathered features during training
+                if self.mkmmd_loss_weight != 0:
+                    self.mkmmd_loss.betas = self.mkmmd_loss.optimize_betas(
+                        X=local_distribution, Y=init_global_distribution, lambda_m=1e-5
+                    )
+                    log(INFO, f"Set optimized betas to minimize distance: {self.mkmmd_loss.betas.squeeze()}.")
+
+        return super().update_after_step(step)
+
+    def update_buffers(
+        self, local_model: torch.nn.Module, init_global_model: torch.nn.Module
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Update the feature buffer of the local and global features."""
+        assert isinstance(local_model, MoonModel)
+        assert isinstance(init_global_model, MoonModel)
+
+        init_state_local_model = local_model.training
+
+        # Set local model to evaluation mode
+        local_model.eval()
+
+        assert not local_model.training
+        assert not init_global_model.training
+
+        local_buffer = []
+        init_global_buffer = []
+
+        with torch.no_grad():
+            for input, target in self.train_loader:
+                input, target = input.to(self.device), target.to(self.device)
+                _, local_features = local_model(input)
+                _, init_global_features = init_global_model(input)
+
+                local_buffer.append(local_features["features"].reshape(len(local_features["features"]), -1))
+                init_global_buffer.append(
+                    init_global_features["features"].reshape(len(init_global_features["features"]), -1)
+                )
+
+        if init_state_local_model:
+            local_model.train()
+
+        return torch.cat(local_buffer, dim=0), torch.cat(init_global_buffer, dim=0)
```

Comment:
A comment that the assumption is that the tensors are "batch-first" so our concatenation is in `dim=0` would be helpful I think.